### ISSUE TITLE
Cannot use search filters with assistive tech or keyboard only

### ISSUE THREAD

**Issue by selfthinker** [splitbrain/dokuwiki#3459]
The search filters cannot be used with keyboard only at all, they can only be used with screen readers if you can see and therefore know what they are and how to use them, and it's more difficult than necessary with voice control tools as well.

This can be a complete barrier but as the functionality of the filter dropdowns is not essential to using the search, this has only a medium priority.

## Using the filters with assistive tech

When using voice control (Dragon) I couldn't click on the filters or interact with them in any way except via the more cumbersome 'mouse grid' fallback. That means I couldn't use the normal way of saying, for example, 'click exact match' to open the 'exact match' dropdown. I had to slowly zone into the spot that I wanted to click on instead (via a grid on the screen which gets smaller and smaller).

Screen readers would usually read out the text and the triangle icon. (Although I suspect older screen readers will probably not read the triangle.) But most would not announce what it is and in which state it would be. The selected item is also just normal text and its functionality is unclear. Understanding how the filters work would be really difficult. They work when you know what to do, though.

NVDA announces that a filter dropdown is "clickable" but not why. It also says that the dropdown has expanded when you get into it.
I made a [video of this interaction in NVDA with Firefox](https://filedn.com/lKLODU2RyEfLPOaqpGeFXVY/dokuwiki/search-filter-dropdowns-nvda.mp4).

VoiceOver announces less than NVDA does. There is nothing to indicate the dropdown text is interactive and that anything has expanded.
Here is a [video of that interaction in VoiceOver with Safari on iOS](https://filedn.com/lKLODU2RyEfLPOaqpGeFXVY/dokuwiki/search-filter-dropdowns-voiceover.mp4) as well.

## Code fails

* [ ] The search filters' role and states cannot be programmatically determined. The trigger each has a generic role (= no role). Each has an `aria-haspopup` which is not linked to the popup and is used on a non-interactive element. The `aria-expanded` on the toggled list is according to one of the two ways described in the spec but is not read out by most screen readers. When navigating to a filter, nothing is announced as to what will happen. When clicking on it, nothing is announced as to what has happened, and there is nothing to indicate that something has opened somewhere.
That fails [4.1.2 Role, Name, Value](https://www.w3.org/TR/WCAG21/#name-role-value).

* [ ] The search filters look like select boxes but they don't behave like any. They are divs and links and don't have the equivalent ARIA roles.
That fails [1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG21/#info-and-relationships).

* [ ] The search filter dropdowns cannot be operated with keyboard only. You tab from the search button straight to the first matching pagename and cannot get to the filters at all.
That fails [2.1.1 Keyboard](https://www.w3.org/TR/WCAG21/#keyboard).
![](https://user-images.githubusercontent.com/108893/113488073-b617c500-94b3-11eb-9d4d-7454313cb625.png)

I would advise against trying to fix the above 3 issues with ARIA but instead rely on existing native HTML elements and therefore follow [the first rule of ARIA](https://www.w3.org/TR/using-aria/#rule1) (= not to use ARIA).
Two elements come to mind, either a `select` or a `details` element. You would get accessibility for free when using those.

## Colour fail

* [ ] The dropdown indicator icons (the grey triangles) are `#999999` against `#ffffff` which have an insufficient colour contrast (2.8:1). The contrast needs to be at least 3:1. Otherwise people who cannot see well might not realise that this is a dropdown as the triangle is the only indicator which shows that it is a dropdown.
That fails [1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast).
